### DISCUSSION TITLE
pg_cron connection failed

### DISCUSSION THREAD

**Discussion by figglewiggle** [supabase/supabase#27763] [Questions] [answered]
I am using pg_cron to schedule a row to be created in a specific table in my database (public schema) every set amount of time. I am able to create the cron job (I have checked the job table in the cron schema), but in the job details, the connection failed for every single attempt it has made. I have enabled permissions for the username that's in the job row, and I can run the function manually in the SQL Editor. Here is the function definition for context that schedules the cron job:

 `DECLARE
  cron_schedule TEXT;
  next_occur TIMESTAMPTZ;
BEGIN
  -- Calculate initial next occurrence
  next_occur := NEW.next_occurence;

  -- Construct the cron schedule based on the next occurrence
  IF NEW.interval_unit = 'hour' THEN
    cron_schedule := date_part('minute', next_occur)::TEXT || ' */' || NEW.interval_value || ' * * *';
  ELSIF NEW.interval_unit = 'day' THEN
    cron_schedule := date_part('minute', next_occur)::TEXT || ' ' || date_part('hour', next_occur)::TEXT || ' */' || NEW.interval_value || ' * *';
  ELSIF NEW.interval_unit = 'week' THEN
    cron_schedule := date_part('minute', next_occur)::TEXT || ' ' || date_part('hour', next_occur)::TEXT || ' * * ' || (date_part('dow', next_occur)::INT % 7);
  ELSIF NEW.interval_unit = 'month' THEN
    cron_schedule := date_part('minute', next_occur)::TEXT || ' ' || date_part('hour', next_occur)::TEXT || ' ' || date_part('day', next_occur)::TEXT || ' */' || NEW.interval_value || ' *';
  ELSE
    RAISE EXCEPTION 'Unsupported recurrence unit: %', NEW.interval_unit;
  END IF;

  -- Schedule the cron job using pg_cron
  PERFORM cron.schedule(
    'create_recurring_task_' || NEW.id,  -- jobname
    cron_schedule,                      -- schedule
    'SELECT create_recurring_task(' || NEW.id || ');'  -- command
  );

  RETURN NEW;
END; `

Additionally, I checked the max_worker_processes and cron.max_running_jobs and found that the amount of max_worker_processes is less than cron.max_running_jobs (6 to 32), so the setup I have is not scalable. 

*Update: I figured out why the connection is failing, but I have no clue how to solve it. The username in the job table is 'anon', and when the cron job is run, it tries to login with 'anon', but 'anon' does not have any login credentials, so the connection fails. How do I make sure I pass an appropriate user to the cron job (e.g. postgres)?

**Comment by GaryAustin1** [comment, marked as answer]
You probably will need to have your function run as `security definer` type so it runs as postgres role.

**Reply by figglewiggle** [reply]
Thank you! I'll let you know how it goes.

**Reply by figglewiggle** [reply]
While we're on the topic, am I able to use security definer to unschedule jobs as well? I've tried to do so and it won't give me permission. It seems that only the supabase_admin can unschedule cron jobs and clear the job detail logs. I've tried granting access to the postgres role through the sql editor, but that does not work, and I don't have a local host instance to work with.

**Reply by GaryAustin1** [reply]
You might try disabling and enabling the cron extension from the extensions UI especially if you have an older project.

**Reply by figglewiggle** [reply]
sounds good thank you!

**Reply by figglewiggle** [reply]
Would I have to redefine the functions I've made using cron or do I just do a regular disable/enable?

**Reply by GaryAustin1** [reply]
Not sure what you mean by redefine.  You definitely will need them to be security definer so they run with a role that can do cron (authenticated and anon can't).   Toggle the extension of and on in dashboard database extensions.

**Reply by figglewiggle** [reply]
Never mind it works just fine now. I just meant that do the functions I defined that were accessing the cron schema need to be deleted and re-entered (which in hindsight makes no sense). Thank you for your help!!!
